### PR TITLE
Add support of Xen PV console via UART-like driver

### LINKED ISSUE #37239: Initial Xen public headers adding
## Origin

Xen Project
Link: [Xen git](https://xenbits.xen.org/gitweb/?p=xen.git;a=summary)
Branch: RELEASE-4.15.0

## Purpose

These headers describe interface to Xen hypercalls, structures, defines etc. They are needed for Xen-related features development. 

## Mode of integration

I would like to place headers to main include directory (`include/xen/public`), they will be used by Xen drivers (PV console, event channels, Xenbus etc.).

## Pull Request

Link to PR: https://github.com/zephyrproject-rtos/zephyr/pull/36691

## Description

As I mentioned, these headers are needed for adding Xen features support to Zephyr project (they provide generic interface for drivers).
In future, during Xen drivers development, the list of required Xen headers may be increased.

# Dependencies

Since it is just public headers, that will be included to Zephyr source tree, no direct dependencies to external component will be created.

## Revision

Modified headers from Xen RELEASE-4.15.0, present in mentioned PR.

## License

``MIT``

### PR DESCRIPTION
**Pull request by firscity**
This commit series adds initial items for Xen features development (some of public headers, hypercall interface etc).

Also it adds minimal implementation of  Xen PV console in Zephyr (via UART-like driver with poll API). It allows to get boot log and application stdout directly from domain hypervisor console, instead of emulated PL011 SBSA.

As addition, this console also supports console_io interface. It can be used for early boot log - driver introduces printk/stdout char-based hooks, that allow to receive kernel messages quickly after start, but require debug version of Xen (compiled with CONFIG_DEBUG option). 

Fixes #37239

### CODE REVIEW

**Review by bertrand-marquis — CHANGES_REQUESTED** [review]
Xen 4.16 is not release yet so if you took current headers status they might change before the 4.16 release in 6 month or so.
I would suggest to use a stable release when importing public headers (4.15.0)

**Review by bertrand-marquis — APPROVED** [review]

**Review by bertrand-marquis — APPROVED** [review]

File: boards/arm64/xenvm/doc/index.rst
Comment (bertrand-marquis) [inline boards/arm64/xenvm/doc/index.rst:80]: This changes looks weird.
Most of the features listed after are under development or not supported at all (xenbus for example).
More then that is is incoherent with the next chapter saying "now only following feature are supported".
Could this change be a mistake ?

**Review by bertrand-marquis — CHANGES_REQUESTED** [review]

**Comment by firscity** [comment]
> Xen 4.16 is not release yet so if you took current headers status they might change before the 4.16 release in 6 month or so.
> I would suggest to use a stable release when importing public headers (4.15.0)

I agree, did not pay enough attention for that. Xen public headers did not pass Zephyr code style and license check, so I decided to fix it altogether with your suggestion. Current headers were taken from Xen RELEASE-4.15.0 branch and modified to comply with Zephyr code-style.

Thank you for comment!

**Review by bertrand-marquis — APPROVED** [review]

**Review by bertrand-marquis — APPROVED** [review]

**Review by bertrand-marquis — APPROVED** [review]

**Review by bertrand-marquis — APPROVED** [review]

File: boards/arm64/xenvm/doc/index.rst
Comment (firscity) [reply boards/arm64/xenvm/doc/index.rst]: Oh, sorry, agree. I missed "not" and context was broken. Of course, these feature are NOT supported yet.

Thank you for careful review!

**Comment by firscity** [comment]
I need someone, who can clarify me few moment with Zephyr CI checks. The only 2 problems left in this PR according to CI report:
1. 2 spacing errors in macro definition :
```
-:852: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
852: FILE: include/xen/public/arch-arm.h:177:
+	typedef union { type *p; unsigned long q; }	\
```
It looks like style checker identify it as multiplication instead of pointer definition and produce spacing error (false-positive).

2. Licensing. Is it allowed to use MIT-licensed headers in Apache 2.0 Zephyr?
CI checker marked such headers as errors:
```
Error: * scan/include/xen/public/arch-arm.h is not apache-2.0 licensed: mit
* scan/include/xen/public/event_channel.h is not apache-2.0 licensed: mit
...
```
But as I can see,  MIT headers/files are already present in Zephyr tree (e.g. include/net/http_parser.h). Is it possible to use original Xen public headers or I should create my own Apache 2.0 licensed files?

File: drivers/serial/Kconfig.xen
Comment (carlocaione) [inline drivers/serial/Kconfig.xen:21]: indentation. Same for the next symbol help. Also s/get/gets/.

File: drivers/serial/uart_hvc_xen.c
Comment (carlocaione) [inline drivers/serial/uart_hvc_xen.c:133]: Can you use `device_map()` here instead?

**Comment by carlocaione** [comment]
One think I don't like is that we now have `arch/arm64/xen` when the proper place for this should be `arch/arm64/core/xen` if we want to follow what all archs are doing.

File: drivers/serial/Kconfig.xen
Comment (firscity) [reply drivers/serial/Kconfig.xen]: Fixed.

**Comment by firscity** [comment]
> One think I don't like is that we now have arch/arm64/xen when the proper place for this should be arch/arm64/core/xen if we want to follow what all archs are doing.

Fixed, hypercall interface moved to arch/arm64/core/xen.

File: drivers/serial/uart_hvc_xen.c
Comment (firscity) [reply drivers/serial/uart_hvc_xen.c]: Yes, you are right, it is better to use device_map() here. I had some problems with it at the start of development, so I tried arch_mem_map() and it worked OK. But now I understand my problem and fixed it, so device_map() started to work OK too.

Also it fixes few other problems at the time (which were created by me and direct usage of arch functions in driver code):
1. It gives virt_addr for ring buffer MMIO (no need to use same addr for source and dest during mapping, so I removed TODO from code).
2. It is accessible directly from main include directory (sys/device_mmio.h), so there is no need to add additional include dirs in CMake.

Thank you for review and great advices!

File: include/xen/public/xen.h
Comment (firscity) [reply include/xen/public/xen.h:32]: These macros are GCC pre-defined for ARM/ARM64, so it works and I decided to not change this part of original Xen header.
But if it matters, I can replace it with CONFIG definition.

File: drivers/serial/uart_hvc_xen.c
Comment (firscity) [reply drivers/serial/uart_hvc_xen.c:117]: I am not familiar with Zephyr logging subsys, so I do not know if LOG output will be available this early. I used printk + early console_io config for debug during development and I am sure, that these logs will be available in case of error.

Probably, to make the same for LOG we need to define Xen console_io logging backend.

File: include/xen/public/xen.h
Comment (povergoing) [reply include/xen/public/xen.h:32]: Personally, emmm, I am just concerned about whether it works for other compilers because Zephyr supports multi compilers.

File: drivers/serial/uart_hvc_xen.c
Comment (povergoing) [reply drivers/serial/uart_hvc_xen.c:117]: Oh, it's ok, as you like :).  But for your reference, I just know if `CONFIG_LOG_MINIMAL ` enabled, LOG_ERR/WRN/INF are the same with printk plus some log level control...

File: include/xen/public/xen.h
Comment (firscity) [reply include/xen/public/xen.h]: Oh, I understood your concerns. Sure, it will be better to use CONFIG-based selection, other compilers have different pre-defined values. Fixed.

Thank you.

File: drivers/serial/uart_hvc_xen.c
Comment (firscity) [reply drivers/serial/uart_hvc_xen.c]: I didn't know about `CONFIG_LOG_MINIMAL`. As I now see LOG_X macros are way more popular(and more functional), then printk in Zephyr, did not pay attention to it.

I replaced printk with logging subsys and confirm, that it works ok with Xen early console_io:
```
root@h3ulcb:~/zephyr# xl create -c zephyr.conf
Parsing config from zephyr.conf
...
(d4) [18692.244316] I: Xen HVC inited successfully
*** Booting Zephyr OS build zephyr-v2.6.0-694-g91b9a5100bc7  ***
```

Thank you for great suggestion! Fixed.

File: drivers/serial/uart_hvc_xen.c
Comment (carlocaione) [inline drivers/serial/uart_hvc_xen.c:23]: empty lines.

File: drivers/serial/uart_hvc_xen.c
Comment (carlocaione) [inline drivers/serial/uart_hvc_xen.c:77]: Coding style:
```suggestion
	if (sent) {
		notify_evtchn(hvc_data->evtchn);
	}
```

**Comment by carlocaione** [comment]
Overall looks good to me, only a couple of nits. About the failing checks:
- The `WARNING:NEW_TYPEDEFS:` are fine but can you please fix the `ERROR:SPACING:`?
- About the license not sure. Maybe @kestewart has a suggestion about that?

File: drivers/serial/uart_hvc_xen.c
Comment (firscity) [reply drivers/serial/uart_hvc_xen.c]: Fixed.

File: drivers/serial/uart_hvc_xen.c
Comment (firscity) [reply drivers/serial/uart_hvc_xen.c]: Fixed.

**Comment by firscity** [comment]
> The WARNING:NEW_TYPEDEFS: are fine but can you please fix the ERROR:SPACING:?

This errors probably are the false-positives, because I don't know what is wrong with such declaration:
```
	typedef union { type *p; unsigned long q; }	\
		__guest_handle_ ## name;		\
```
It looks like checking script recognizes this "type \*p" declaration in typedef as multiplication of "type\*p" and produce this error:
```
-:852: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
852: FILE: include/xen/public/arch-arm.h:177:
+	typedef union { type *p; unsigned long q; }	\

```
 But it just a pointer declaration inside union with type, which will be passed to macro. AFAIK, such coding style is OK.

**Comment by carlocaione** [comment]
LGTM but I have no idea for the license problem.

**Review by carlocaione — APPROVED** [review]

**Comment by povergoing** [comment]
> > The WARNING:NEW_TYPEDEFS: are fine but can you please fix the ERROR:SPACING:?
> 
> This errors probably are the false-positives, because I don't know what is wrong with such declaration:
> 
> ```
> 	typedef union { type *p; unsigned long q; }	\
> 		__guest_handle_ ## name;		\
> ```
> 
> It looks like checking script recognizes this "type *p" declaration in typedef as multiplication of "type*p" and produce this error:
> 
> ```
> -:852: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
> 852: FILE: include/xen/public/arch-arm.h:177:
> +	typedef union { type *p; unsigned long q; }	\
> ```
> 
> But it just a pointer declaration inside union with type, which will be passed to macro. AFAIK, such coding style is OK.

Would you like to try `type* p`?

**Comment by firscity** [comment]
> Would you like to try type* p?

I tried checkpatch.pl locally - result is the same. The only "type * p" works OK, it confirms that checkpatch detects this as multiplication. Thus, I think that this is not a valid checkpatch error (false-positive).

**Review by povergoing — APPROVED** [review]
Look good to me. But I don't know the Licence issue too.

**Comment by dcpleung** [comment]
Here is the doc about pulling in external components and info on licensing: https://docs.zephyrproject.org/latest/contribute/external.html#external-contributions

tl;dr: probably needs TSC to approve.

**Comment by firscity** [comment]
Changes were rebased on top of `zephyrproject-rtos:main`.
Conflicts in `CODEOWNERS` and `drivers/serial/Kconfig` are now resolved.

**Comment by firscity** [comment]
Dear @carlocaione and @dcpleung, can you please clarify: should I do any actions to make TSC to process this pull-request or I should just wait for decision?

**Comment by dcpleung** [comment]
I think first step is to send a quick email to TSC mailing list about the licensing. Hopefully it would be easy since it's BSD.

**Comment by carlescufi** [comment]
@firscity sorry for the delay. I've included this in this week's TSC agenda.

**Comment by firscity** [comment]
@carlescufi, thank you a lot for help, I will wait for results. In case of TSC approval, I will prepare this PR for merge ASAP (fix merge conflicts).

**Comment by cfriedt** [comment]
@firscity - please rebase

**Comment by firscity** [comment]
@cfriedt, done.

**Review by dcpleung — APPROVED** [review]

**Comment by cfriedt** [comment]
@firscity - the TSC has decided to hold off merging this, but I would imagine it will go in in the near future. Thanks for your patience.

**Comment by firscity** [comment]
In 2.7 release `CONFIG_UART_CONSOLE_ON_DEV_NAME` Kconfig option was removed, so it was needed to change this PR a bit. This option was removed from `xenvm_defconfig` and console selection was moved to `chosen` device-tree node.

No functional changes were made by last force-push.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
